libfwupdplugin/fu-bios-attrs.c

```diff
@@ -0,0 +1,424 @@
+		if (driver_dir == NULL)
+			return FALSE;
+		do {
+			const gchar *name = g_dir_read_name(driver_dir);
+			g_autoptr(GError) error_local = NULL;
+			if (name == NULL)
+				break;
+			if (!fu_bios_attrs_populate_attribute(self,
+							      driver,
+							      path,
+							      name,
+							      &error_local)) {
+				if (g_error_matches(error_local,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_NOT_SUPPORTED)) {
+					g_debug("%s is not supported", name);
+					continue;
+				}
+				g_propagate_error(error, g_steal_pointer(&error_local));
+				return FALSE;
+			}
+		} while (TRUE);
+	} while (TRUE);
+
+	return TRUE;
+}
+
+static void
+fu_bios_attrs_init(FuBiosAttrs *self)
+{
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+	self->attrs = g_ptr_array_new_with_free_func((GDestroyNotify)g_object_unref);
+}
+
+/**
+ * fu_bios_attributes_get_attr:
+ * @self: a #FuBiosAttrs
+ * @name: the attribute "name" to check for
+ *
+ * Returns: (transfer none): the attribute with the given name or NULL if it doesn't exist.
+ *
+ * Since: 1.8.4
+ **/
+FwupdBiosAttr *
+fu_bios_attributes_get_attr(FuBiosAttrs *self, const gchar *name)
+{
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), NULL);
+	g_return_val_if_fail(name != NULL, NULL);
+
+	for (guint i = 0; i < self->attrs->len; i++) {
+		FwupdBiosAttr *attr = g_ptr_array_index(self->attrs, i);
+		const gchar *tmp = fwupd_bios_attr_get_name(attr);
+		if (g_strcmp0(tmp, name) == 0)
+			return attr;
+	}
+	return NULL;
+}
+
+/**
+ * fu_bios_attrs_get_all:
+ * @self: a #FuBiosAttrs
+ *
+ * Gets all the attributes in the object.
+ *
+ * Returns: (transfer container) (element-type FwupdBiosAttr): attributes
+ *
+ * Since: 1.8.4
+ **/
+GPtrArray *
+fu_bios_attrs_get_all(FuBiosAttrs *self)
+{
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), NULL);
+	return g_ptr_array_ref(self->attrs);
+}
+
+/**
+ * fu_bios_attributes_get_pending:
+ * @self: a #FuBiosAttrs
+ * @result: (out): Whether a reboot is pending
+ * @error: (nullable): optional return location for an error
+ *
+ * Determines if the system will apply changes to attributes upon reboot
+ *
+ * Since: 1.8.4
+ **/
+gboolean
+fu_bios_attributes_get_pending(FuBiosAttrs *self, gboolean *result, GError **error)
+{
+	FwupdBiosAttr *attr;
+	const gchar *data;
+	guint64 val = 0;
+
+	g_return_val_if_fail(result != NULL, FALSE);
+	g_return_val_if_fail(FU_IS_BIOS_ATTRS(self), FALSE);
+
+	for (guint i = 0; i < self->attrs->len; i++) {
+		const gchar *tmp;
+
+		attr = g_ptr_array_index(self->attrs, i);
+		tmp = fwupd_bios_attr_get_name(attr);
+		if (g_strcmp0(tmp, FWUPD_BIOS_ATTR_PENDING_REBOOT) == 0)
+			break;
+		attr = NULL;
+	}
+
+	if (attr == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_NOT_FOUND,
+				    "failed to find pending reboot attribute");
+		return FALSE;
+	}
+
+	data = fwupd_bios_attr_get_current_value(attr, error);
+	if (!fu_strtoull(data, &val, 0, G_MAXUINT32, error))
+		return FALSE;
+
+	*result = (val == 1);
```

Comment:
i.e. if (result != NULL) *result =

Reply:
I think that's a programming error though.  This function's purpose is to return an output.
`	g_return_val_if_fail(result != NULL, FALSE);`

Reply:
Ya, fair enough, point taken.